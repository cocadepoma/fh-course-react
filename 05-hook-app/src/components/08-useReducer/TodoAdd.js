import React from "react";
import { useForm } from "../../hooks/useForm";

export const TodoAdd = ({ handleAddTodo }) => {
    const [{ description }, handleInputChange, reset] = useForm({
        description: "",
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if (description.trim().length <= 1) {
            return;
        }

        handleAddTodo({
            id: new Date().getTime(),
            desc: description,
            done: false,
        });

        // reset();
    };
    return (
        <>
            <h4>Agregar TODO</h4>
            <hr />

            <form onSubmit={handleSubmit} className='d-flex flex-column'>
                <input
                    onChange={handleInputChange}
                    className='form-control'
                    type='text'
                    name='description'
                    value={description}
                    placeholder='Aprender...'
                    autoComplete='off'
                />

                <button type='submit' className='btn btn-block btn-outline-primary mt-1'>
                    Agregar
                </button>
            </form>
        </>
    );
};
