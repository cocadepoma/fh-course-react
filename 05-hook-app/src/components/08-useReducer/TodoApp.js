import React, { useEffect, useReducer } from "react";
import { todoReducer } from "./todoReducer";
import { TodoList } from "./TodoList";
import { TodoAdd } from "./TodoAdd";
import "./styles.css";

export const TodoApp = () => {
    // TODOAPP LIST

    const init = () => {
        return JSON.parse(localStorage.getItem("todos")) || [];
        // return [
        //     { id: new Date().getTime(), desc: "Aprender React", done: false },
        // ];
    };

    const [todos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem("todos", JSON.stringify(todos));
    }, [todos]);

    const handleDelete = (id) => {
        const action = {
            type: "delete",
            payload: id,
        };
        dispatch(action);
    };

    const handleToggle = (id) => {
        dispatch({
            type: "toggle",
            payload: id,
        });
    };

    const handleAddTodo = (desc) => {
        dispatch({
            type: "add",
            payload: desc,
        });
    };
    return (
        <div>
            <h1>Todo APP ({todos.length})</h1>
            <hr />
            <div className='row'>
                <div className='col-md-7'>
                    <TodoList todos={todos} handleToggle={handleToggle} handleDelete={handleDelete} />
                </div>
                <div className='col-md-5'>
                    <TodoAdd handleAddTodo={handleAddTodo} />
                </div>
            </div>
        </div>
    );
};
