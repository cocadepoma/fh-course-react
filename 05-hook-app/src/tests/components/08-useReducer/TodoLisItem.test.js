import { TodoListItem } from "../../../components/08-useReducer/TodoListItem";
import { shallow } from "enzyme";
import { demoTodos } from "../../fixtures/demoTodos";

describe("Pruebas en <TodoLisItem/>", () => {
    const todo = demoTodos[0];

    const handleToggle = jest.fn();
    const handleDelete = jest.fn();
    const index = 0;
    const wrapper = shallow(
        <TodoListItem todo={todo} index={index} handleToggle={handleToggle} handleDelete={handleDelete} />
    );

    test("debe de mostrarse corectamente", () => {
        //Snapshot
        expect(wrapper).toMatchSnapshot();
    });

    test("debe de llamar la función handleDelete", () => {
        wrapper.find("p").simulate("click");
        expect(handleToggle).toHaveBeenCalledWith(todo.id);
        expect(handleDelete).not.toHaveBeenCalled();
    });
    test("debede llamar la función handleToggle", () => {
        wrapper.find("button").simulate("click");
        expect(handleDelete).toHaveBeenCalledWith(todo.id);
        expect(handleToggle).not.toHaveBeenCalled();
    });
    test("debe de mostrar el texto correctamente", () => {
        expect(wrapper.find("p").text().trim()).toBe(`${index + 1} - ${todo.desc}`);
    });
    test("debe de tener la clase complete, si el todo.done está en true", () => {
        todo.done = true;
        const wrapper = shallow(
            <TodoListItem todo={todo} index={index} handleToggle={handleToggle} handleDelete={handleDelete} />
        );
        expect(wrapper.find("p").hasClass("complete")).toBe(true);
    });
});
