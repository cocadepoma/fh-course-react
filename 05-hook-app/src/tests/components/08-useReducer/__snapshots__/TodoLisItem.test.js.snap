// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoLisItem/> debe de mostrarse corectamente 1`] = `
<li
  className="list-group-item"
  key="1"
>
  <p
    className="false"
    onClick={[Function]}
  >
    1
     - 
    Aprender React
  </p>
  <button
    className="btn btn-danger me-5"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
